### ISSUE TITLE
SPARQL: extra indentation after PrefixedName followed by dot without space

### ISSUE THREAD

**Issue by lucaswerkmeister** [codemirror/CodeMirror#6637]
Since 9caacec1900d71a971561147ba1e8acb2f08609c (#6440 – CC @MarkBoyes, @marijnh), SPARQL mode suggests the following indentation:

```sparql
ASK {
  ?a ?b c:.
     # ?
}
```

This seems to happen whenever the object of a triple is a prefixed name, and there is no space between the object and the dot terminating the triple. The following all work better:

```sparql
ASK {
  ?a ?b c: .
  # !
}
```

```sparql
ASK {
  ?a ?b ?c.
  # !
}
```

```sparql
ASK {
  ?a ?b <c>.
  # !
}
```

The issue can be readily tested in the `mode/sparql/index.html` that’s part of CodeMirror.

I can reproduce this in Firefox (87) and Chromium (89), and in CodeMirror releases 5.59.2 and 5.60.0 as well as current master (5895973d86f60b8e5d8fd1aebb4cca70b8930621).

Also tracked at Wikimedia as [T278685](https://phabricator.wikimedia.org/T278685).

**Comment by marijnh** [comment]
It's been ages since I used Spaql and the spec if a hard-to-read web of links to other specs, so could you summarize for me what the `c:.` syntax is intended to mean?

**Comment by lucaswerkmeister** [comment]
`c:` is perhaps an overly terse example. `schema:Person`, for instance, is equivalent to the URI http://schema.org/Person (which you could also specify in SPARQL as `<http://schema.org/Person>`, in angle brackets), assuming that a `PREFIX schema: <http://schema.org/>` was already declared in the query. `c:` would be a valid prefix name (with an empty local part) if a `PREFIX c:` was declared. (But I wouldn’t expect CodeMirror to check if a prefix is defined or not.)

**Comment by MarkBoyes** [comment]
I concur: with my colour scheme its is obvious what is happening: If the dot is placed immediately after the colon is is being parsed as a local part of the IRI and id displayed in the same colour as the prefix part, but is a different colour if a space is added.

'.' is a legal character in PN_LOCAL but not as the first character: the function eatPnLocal I added does not account for that.

**Comment by lucaswerkmeister** [comment]
I should perhaps clarify that this issue also affects nonempty local parts. And if I read the SPARQL grammar correctly, `.` is also not a legal character at the end of `PN_LOCAL`.

**Comment by lucaswerkmeister** [comment]
Great, thanks!
